presto-main/src/main/java/com/facebook/presto/execution/TrackingRemoteTaskFactory.java

```diff
@@ -87,6 +91,15 @@ public synchronized void stateChanged(TaskStatus newStatus)
             previousUserMemory = currentUserMemory;
             previousSystemMemory = currentSystemMemory;
             stateMachine.updateMemoryUsage(deltaUserMemoryInBytes, deltaTotalMemoryInBytes, currentUserMemory, currentTotalMemory);
+
+            if (state == PLANNED && newStatus.getState() == RUNNING) {
+                int runningTasks = stateMachine.getCurrentRunningTasks().incrementAndGet();
+                stateMachine.getPeakRunningTasks().accumulateAndGet(runningTasks, Math::max);
+            }
+            else if (!state.isDone() && newStatus.getState().isDone()) {
+                stateMachine.getCurrentRunningTasks().decrementAndGet();
```

Comment:
Can this go negative on some state transitions such as when it goes from PLANNED to anything other then  RUNNING?

Reply:
Very good point, updated this logic to be safe